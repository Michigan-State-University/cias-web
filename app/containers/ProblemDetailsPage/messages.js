/*
 * HomePage Messages
 *
 * This contains all the text for the HomePage container.
 */
import { defineMessages } from 'react-intl';

export const scope = 'app.containers.HomePage';

export default defineMessages({
  createIntervention: {
    id: `${scope}.createIntervention`,
    defaultMessage: 'Create intervention',
  },
  noInterventions: {
    id: `${scope}.noInterventions`,
    defaultMessage: `You don't have any interventions`,
  },
  filter: {
    id: `${scope}.filter`,
    defaultMessage: `Filter by name`,
  },
  noFilterResults: {
    id: `${scope}.noFilterResults`,
    defaultMessage: `There are no interventions with this criteria`,
  },
  showArchived: {
    id: `${scope}.showArchived`,
    defaultMessage: `Show Archived`,
  },
  back: {
    id: `${scope}.back`,
    defaultMessage: 'Back to dashboard',
  },
  placeholder: {
    id: `${scope}.placeholder`,
    defaultMessage: 'e-Intervention Name',
  },
  adjust: {
    id: `${scope}.adjust`,
    defaultMessage: 'Edit',
  },
  copy: {
    id: `${scope}.copy`,
    defaultMessage: 'Send copy to researcher',
  },
  duplicate: {
    id: `${scope}.duplicate`,
    defaultMessage: 'Duplicate intervention',
  },
  archive: {
    id: `${scope}.archive`,
    defaultMessage: 'Archive intervention',
  },
  modalTitle: {
    id: `${scope}.modalTitle`,
    defaultMessage: 'Choose researchers to send',
  },
  participantShareModalTitle: {
    id: `${scope}.participantShareModalTitle`,
    defaultMessage: 'Send an email with the invitation',
  },
  anyone: {
    id: `${scope}.anyone`,
    defaultMessage: 'Anyone with the link can access sessions',
  },
  registered: {
    id: `${scope}.registered`,
    defaultMessage:
      'Anyone who is a registered participant can access sessions',
  },
  invited: {
    id: `${scope}.invited`,
    defaultMessage: 'Only selected registered participants can access sessions',
  },
});
