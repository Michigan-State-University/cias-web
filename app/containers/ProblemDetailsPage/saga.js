import { all } from 'redux-saga/effects';
import {
  copyInterventionSaga,
  editProblemSaga,
  fetchProblemSaga,
  sendProblemCsvSaga,
  updateInterventionSettingsSaga,
  reorderSessionsSaga,
  createInterventionSaga,
} from 'global/reducers/problem/sagas';
import { copyProblemSaga } from 'global/reducers/problems/sagas';

export default function* problemDetailsPageSagas() {
  yield all([
    createInterventionSaga(),
    fetchProblemSaga(),
    editProblemSaga(),
    sendProblemCsvSaga(),
    copyInterventionSaga(),
    copyProblemSaga(),
    updateInterventionSettingsSaga(),
    reorderSessionsSaga(),
  ]);
}
