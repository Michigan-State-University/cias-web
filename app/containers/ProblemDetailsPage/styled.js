import styled from 'styled-components';
import { themeColors, colors, boxShadows } from 'theme';

export const StatusLabel = styled.button`
  border: none;
  display: flex;
  padding: 5px 10px;
  font-size: 14px;
  font-weight: bold;
  line-height: 16px;
  border-radius: 5px;
  background: ${props =>
    props.active ? themeColors.primary : props.color || themeColors.highlight};
  color: ${props => props.active && colors.white};
  box-shadow: ${boxShadows.selago};
  transition: 0.3s;
  &:hover {
    cursor: pointer;
    background: ${themeColors.primary};
    color: ${colors.white};
  }
  &:focus {
    outline: none;
    outline-offset: 0px;
  }
`;
