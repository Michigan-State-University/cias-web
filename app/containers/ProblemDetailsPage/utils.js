export const updateStatuses = {
  draft: 'broadcast',
  published: 'close',
};
