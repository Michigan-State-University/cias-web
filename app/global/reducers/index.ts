import { liveChatReducerKey } from './liveChat/reducer';
import { LiveChatState } from './liveChat/types';
import { chatWidgetReducerKey } from './chatWidget/reducer';
import { ChatWidgetState } from './chatWidget/types';
import { navigatorSetupReducerKey } from './navigatorSetup/reducer';
import { NavigatorSetupState } from './navigatorSetup/types';
import { notificationsReducerKey } from './notifications/reducer';
import { NotificationsState } from './notifications/types';

export type RootState = {
  [liveChatReducerKey]: LiveChatState;
  [chatWidgetReducerKey]: ChatWidgetState;
  [navigatorSetupReducerKey]: NavigatorSetupState;
  [notificationsReducerKey]: NotificationsState;
};
