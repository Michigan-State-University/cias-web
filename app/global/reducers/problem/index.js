export { CREATE_PROBLEM_SUCCESS } from './constants';
export {
  fetchProblemRequest,
  createProblemRequest,
  editProblemRequest,
} from './actions';
export { problemReducer } from './reducer';
export {
  makeSelectProblem,
  makeSelectProblemLoader,
  makeSelectProblemState,
} from './selectors';
export { createProblemSaga, editProblemSaga, fetchProblemSaga } from './sagas';
