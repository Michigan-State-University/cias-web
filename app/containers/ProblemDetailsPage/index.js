/*
 *
 * ProblemDetailsPage
 *
 */

import React, { useLayoutEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage, injectIntl } from 'react-intl';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { createStructuredSelector } from 'reselect';
import { Link } from 'react-router-dom';
import get from 'lodash/get';
import Reorder, { reorder } from 'react-reorder';
import orderBy from 'lodash/orderBy';

import { StyledInput } from 'components/Input/StyledInput';
import Loader from 'components/Loader';
import Column from 'components/Column';
import ErrorAlert from 'components/ErrorAlert';
import Row from 'components/Row';
import Box from 'components/Box';
import BackButton from 'components/BackButton';
import ShareBox from 'containers/ShareBox';
import TextButton from 'components/Button/TextButton';
import Text from 'components/Text';
import Dropdown from 'components/Dropdown';
import Modal from 'components/Modal';
import {
  fetchProblemRequest,
  makeSelectProblemState,
  editProblemRequest,
  problemReducer,
  sendProblemCsvRequest,
  reorderInterventionList,
  copyInterventionRequest,
  createInterventionRequest,
  makeSelectCurrentInterventionIndex,
  changeCurrentIntervention,
} from 'global/reducers/problem';
import injectSaga, { useInjectSaga } from 'utils/injectSaga';
import { useInjectReducer } from 'utils/injectReducer';
import { colors, themeColors } from 'theme';
import appStages from 'global/appStages';
import globalMessages from 'global/i18n/globalMessages';

import fileShare from 'assets/svg/file-share.svg';
import copy from 'assets/svg/copy.svg';
import archive from 'assets/svg/archive.svg';
import { closed } from 'models/Status/StatusTypes';
import { copyProblemRequest } from 'global/reducers/problems';

import {
  getQuestionsRequest,
  questionsReducer,
  getQuestionsSaga,
} from 'global/reducers/questions';

import Spinner from 'components/Spinner';
import { StatusLabel, InterventionOptions } from './styled';
import problemDetailsPageSagas from './saga';
import InterventionCreateButton from './components/InterventionCreateButton';
import InterventionStatusButtons from './components/InterventionStatusButtons';
import InterventionListItem from './components/InterventionListItem';
import SelectResearchers from './components/SelectResearchers';
import messages from './messages';
import { updateStatuses } from './utils';

export function ProblemDetailsPage({
  intl: { formatMessage },
  createIntervention,
  editProblem,
  fetchProblem,
  match: {
    params: { problemId },
  },
  problemState: {
    problem,
    loaders: { fetchProblemLoading, createInterventionLoading },
    errors: { fetchProblemError, createInterventionError },
  },
  interventionIndex,
  changeInterventionIndex,
  sendCsv,
  copyIntervention,
  reorderInterventions,
  copyProblem,
  fetchQuestions,
}) {
  useInjectReducer({
    key: 'problem',
    reducer: problemReducer,
  });
  useInjectReducer({ key: 'questions', reducer: questionsReducer });
  useInjectSaga({ key: 'getQuestions', saga: getQuestionsSaga });

  const { interventions, name, id, status, shared_to: sharedTo } =
    problem || {};

  const [modalVisible, setModalVisible] = useState(false);

  const closeModal = () => setModalVisible(false);
  const openModal = () => setModalVisible(true);
  const handleCopyProblem = () => copyProblem({ problemId: id });
  const handleArchiveProblem = () =>
    editProblem({
      path: 'status_event',
      value: 'to_archive',
    });

  const options = [
    {
      id: 'copy',
      label: formatMessage(messages.copy),
      icon: fileShare,
      action: openModal,
      color: colors.bluewood,
    },
    {
      id: 'duplicate',
      label: formatMessage(messages.duplicate),
      icon: copy,
      action: handleCopyProblem,
      color: colors.bluewood,
    },
    {
      id: 'archive',
      label: formatMessage(messages.archive),
      icon: archive,
      action: handleArchiveProblem,

      color: colors.bluewood,
    },
  ];

  useLayoutEffect(() => {
    fetchProblem(problemId);
  }, []);

  const availableOptions = options.filter(
    elem => elem.id !== 'archive' || status === closed,
  );

  const handleCopyIntervention = interventionId => {
    copyIntervention({ interventionId });
  };

  const editName = val => editProblem({ path: 'name', value: val });

  const handleChangeStatus = () =>
    editProblem({
      path: 'status_event',
      value: get(updateStatuses, status, ''),
    });

  const handleSendCsv = () => sendCsv(id);

  const createInterventionCall = () =>
    createIntervention(problemId, interventions.length);

  const handleReorder = (event, previousIndex, nextIndex) => {
    const newList = reorder(interventions, previousIndex, nextIndex);
    let position = 0;
    const orderdedNewList = newList.map(question => {
      position += 1;
      return {
        ...question,
        position,
      };
    });
    reorderInterventions({
      reorderedList: orderdedNewList,
      problemId,
    });
  };

  const renderList = () => (
    <>
      <Reorder
        reorderId="problem-list"
        onReorder={handleReorder}
        holdTime={125}
      >
        {interventions &&
          orderBy(interventions, 'position').map((intervention, index) => {
            const handleClick = () => {
              if (intervention.position !== interventionIndex + 1) {
                fetchQuestions(intervention.id);
                changeInterventionIndex(index);
              }
            };
            const nextIntervention = interventions.find(
              ({ position }) => position === intervention.position + 1,
            );
            return (
              <Row key={intervention.id}>
                <InterventionListItem
                  intervention={intervention}
                  index={index}
                  isSelected={index === interventionIndex}
                  handleClick={handleClick}
                  handleCopyIntervention={handleCopyIntervention}
                  nextInterventionName={
                    nextIntervention ? nextIntervention.name : null
                  }
                />
              </Row>
            );
          })}
      </Reorder>
    </>
  );

  if (fetchProblemLoading) return <Loader />;

  if (fetchProblemError)
    return <ErrorAlert errorText={fetchProblemError} fullPage />;

  return (
    <Box height="100%" width="100%" padding="60px 160px">
      <Modal
        title={formatMessage(messages.modalTitle)}
        onClose={closeModal}
        visible={modalVisible}
      >
        <SelectResearchers problemId={id} onClose={closeModal} />
      </Modal>

      <Row justify="between">
        <BackButton to="/">
          <FormattedMessage {...messages.back} />
        </BackButton>
      </Row>
      <Row my={18} justify="between">
        <Row align="center">
          <Box mr={15}>
            <StatusLabel status={status}>
              {status && formatMessage(globalMessages.statuses[status])}
            </StatusLabel>
          </Box>
          <StyledInput
            ml={-12}
            px={12}
            py={6}
            width="400px"
            value={name}
            fontSize={23}
            placeholder={formatMessage(messages.placeholder)}
            onBlur={editName}
            maxWidth="none"
          />
        </Row>
        <Row>
          <InterventionStatusButtons
            status={status}
            handleChangeStatus={handleChangeStatus}
            handleSendCsv={handleSendCsv}
          />
          <InterventionOptions>
            <Dropdown options={availableOptions} clickable />
          </InterventionOptions>
        </Row>
      </Row>
      <Row
        bg={colors.linkWater}
        borderRadius={10}
        py={15}
        px={20}
        align="center"
        width="fit-content"
      >
        <Text fontWeight="bold" mr={12}>
          {sharedTo && <FormattedMessage {...messages[sharedTo]} />}
        </Text>
        <Link to={`/interventions/${id}/settings`}>
          <TextButton>
            <FormattedMessage {...messages.adjust} />
          </TextButton>
        </Link>
      </Row>
      <Row>
        <Column sm={6}>
          {renderList()}
          {createInterventionLoading && (
            <Row my={18} align="center">
              <Spinner color={themeColors.secondary} />
            </Row>
          )}
          <Row my={18} align="center">
            <InterventionCreateButton handleClick={createInterventionCall} />
          </Row>
        </Column>
        {createInterventionError && (
          <ErrorAlert errorText={createInterventionError} />
        )}
        {process.env.APP_STAGE === appStages.dev.id && (
          <Column ml={38} sm={6} mt={18}>
            <Column position="sticky" top="100px">
              <ShareBox />
            </Column>
            <div />
          </Column>
        )}
      </Row>
    </Box>
  );
}

ProblemDetailsPage.propTypes = {
  intl: PropTypes.object,
  createIntervention: PropTypes.func,
  fetchProblem: PropTypes.func,
  problemState: PropTypes.shape({
    interventions: PropTypes.array,
    fetchProblemError: PropTypes.string,
    fetchProblemLoading: PropTypes.bool,
  }),
  match: PropTypes.object,
  editProblem: PropTypes.func,
  interventionIndex: PropTypes.number,
  changeInterventionIndex: PropTypes.func,
  sendCsv: PropTypes.func,
  copyIntervention: PropTypes.func,
  reorderInterventions: PropTypes.func,
  copyProblem: PropTypes.func,
  fetchQuestions: PropTypes.func,
};

const mapStateToProps = createStructuredSelector({
  problemState: makeSelectProblemState(),
  interventionIndex: makeSelectCurrentInterventionIndex(),
});

const mapDispatchToProps = {
  createIntervention: createInterventionRequest,
  fetchQuestions: getQuestionsRequest,
  fetchProblem: fetchProblemRequest,
  editProblem: editProblemRequest,
  changeInterventionIndex: changeCurrentIntervention,
  sendCsv: sendProblemCsvRequest,
  copyIntervention: copyInterventionRequest,
  reorderInterventions: reorderInterventionList,
  copyProblem: copyProblemRequest,
};

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps,
);

const withSaga = injectSaga({
  key: 'problemDetailsPageSagas',
  saga: problemDetailsPageSagas,
});

export default compose(
  withConnect,
  withSaga,
  injectIntl,
)(ProblemDetailsPage);
