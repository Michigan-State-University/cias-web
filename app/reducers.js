import { connectRouter } from 'connected-react-router';
import { combineReducers } from 'redux';

import history from 'utils/history';
import languageProviderReducer from 'containers/LanguageProvider/reducer';

import { authReducer } from 'global/reducers/auth';
import { globalStateReducer } from 'global/reducers/globalState';

/**
 * Merges the main reducer with the router state and dynamically injected reducers
 */
export default function createReducer(injectedReducers = {}) {
  const rootReducer = combineReducers({
    auth: authReducer,
    language: languageProviderReducer,
    router: connectRouter(history),
    global: globalStateReducer,
    ...injectedReducers,
  });

  return rootReducer;
}
