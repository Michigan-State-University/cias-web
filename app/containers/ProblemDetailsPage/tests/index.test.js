import React from 'react';
import { render } from 'react-testing-library';
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import { MemoryRouter } from 'react-router-dom';
import { problemReducer } from 'global/reducers/problem';

import ProblemDetailsPage from '../index';
import { DEFAULT_LOCALE } from '../../../i18n';

const defaultProps = {
  match: {
    params: {
      problemId: '1jkajsd2-112a',
    },
  },
};

describe('<ProblemDetailsPage />', () => {
  let store;

  const initialState = {};
  const reducer = state => state;
  beforeAll(() => {
    store = createStore(reducer, initialState);
    store.runSaga = () => {};
    store.injectedReducers = {
      problem: problemReducer,
    };
    store.injectedSagas = {};
  });

  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(
      <Provider store={store}>
        <IntlProvider locale={DEFAULT_LOCALE}>
          <MemoryRouter>
            <ProblemDetailsPage {...defaultProps} />
          </MemoryRouter>
        </IntlProvider>
      </Provider>,
    );
    expect(spy).not.toHaveBeenCalled();
  });

  it.only('Should render loader and match the snapshot', () => {
    const { container } = render(
      <IntlProvider locale={DEFAULT_LOCALE}>
        <Provider store={store}>
          <MemoryRouter>
            <ProblemDetailsPage {...defaultProps} />
          </MemoryRouter>
        </Provider>
      </IntlProvider>,
    );
    expect(container).toMatchSnapshot();
  });
});
