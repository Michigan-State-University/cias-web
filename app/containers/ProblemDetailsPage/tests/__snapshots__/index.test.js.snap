// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProblemDetailsPage /> Should render loader and match the snapshot 1`] = `
<div>
  <div
    class="styled__FillScreenLoader-tle2yi-0 fDUeBu"
    data-testid="fullscreen-loader"
    opacity="0.5"
  >
    <div
      class="styled__StyledSpinner-d123q3-0 hXQJFy"
      color="#C866EA"
      size="100"
      width="6"
    />
  </div>
</div>
`;
